tools/kit_builder.py

```
@@ -1,16 +1,201 @@
+    tokens = notation.split()
+    if not tokens or len(tokens) > MAX_STEPS_PER_TRACK:
+        raise KitError(f"{what}: a template needs 1–{MAX_STEPS_PER_TRACK} steps, got {notation!r}")
+    return [(0, 0) if token == REST_TOKEN else (1, 0) for token in tokens]
+
+
+def degree_list(values, what):
+    if not 1 <= len(values) <= MAX_NOTE_SEQUENCE_LENGTH:
+        raise KitError(f"{what} needs 1–{MAX_NOTE_SEQUENCE_LENGTH} degrees, got {values}")
+    for degree in values:
+        if not isinstance(degree, int) or degree < 0:
+            raise KitError(f"{what}: degrees are non-negative integers, got {degree!r}")
```

Ruff (0.16.3)

[warning] Avoid specifying long messages outside the exception class

(TRY003)